### PR TITLE
Docker buildx support

### PR DESCRIPTION
**Pull request by vivek-devtron**
# Description

Provide user support to choose target  build platform via Docker buildx support at the time of app configuration

Fixes # [(issue)](https://github.com/devtron-labs/dashboard/issues/436)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Manually tested

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/ciConfig/CIConfig.tsx [hunk 1]
```diff
@@ -124,40 +126,65 @@ function Form({ dockerRegistries, sourceConfig, ciConfig, reload, appId }) {
...
         },
         onValidation,
-    );
-    const [args, setArgs] = useState([]);
-    const [loading, setLoading] = useState(false);
-
+    )
+    const [args, setArgs] = useState([])
+    const [loading, setLoading] = useState(false)
+    const [platformError, setPlatformError] = useState<string>('')
```
Comment (shivani170) [inline src/components/ciConfig/CIConfig.tsx:132]: remove (' ')

File: src/components/ciConfig/CIConfig.tsx [hunk 2]
```diff
@@ -124,40 +126,65 @@ function Form({ dockerRegistries, sourceConfig, ciConfig, reload, appId }) {
...
+    const targetPlatformList: OptionType[] = [
+        { label: 'linux/arm64', value: 'linux/arm64' },
+        { label: 'linux/amd64', value: 'linux/amd64' },
+        { label: 'linux/arm/v7', value: 'linux/arm/v7' },
+    ]
+    let _selctedPlatforms = []
```
Comment (shivani170) [inline src/components/ciConfig/CIConfig.tsx:138]: typo error 'selectedPlatforms'

File: src/components/ciConfig/CIConfig.tsx [hunk 3]
```diff
@@ -124,40 +126,65 @@ function Form({ dockerRegistries, sourceConfig, ciConfig, reload, appId }) {
...
+    if (ciConfig && ciConfig.dockerBuildConfig && ciConfig.dockerBuildConfig.targetPlatform) {
```
Comment (shivani170) [inline src/components/ciConfig/CIConfig.tsx:139]: We can use ciConfig?.dockerBuildConfig instead

File: src/components/ciConfig/CIConfig.tsx [hunk 4]
```diff
@@ -124,40 +126,65 @@ function Form({ dockerRegistries, sourceConfig, ciConfig, reload, appId }) {
...
+        _selctedPlatforms = ciConfig.dockerBuildConfig.targetPlatform.split(',').map((platformValue) => {
```
Comment (shivani170) [inline src/components/ciConfig/CIConfig.tsx:140]: Same typo

File: src/components/ciConfig/CIConfig.tsx [hunk 5]
```diff
@@ -124,40 +126,65 @@ function Form({ dockerRegistries, sourceConfig, ciConfig, reload, appId }) {
...
+        })
+        const areArgsWrong = args2.some((arg) => arg.keyError || arg.valueError)
         if (areArgsWrong) {
-            setArgs([...args2]);
-            return;
+            setArgs([...args2])
+            return
+        }
+        let targetPlatforms = ''
+        let set = new Set()
```
Comment (sohel-rp) [inline src/components/ciConfig/CIConfig.tsx:177]: use const and rename the var to something like `targetPlatforms` or similiar

File: src/components/ciConfig/CIConfig.tsx [hunk 6]
```diff
@@ -303,10 +331,101 @@ function Form({ dockerRegistries, sourceConfig, ciConfig, reload, appId }) {
...
+            <components.MenuList {...props}>
+                <div className="cn-5 pl-12 pt-4 pb-4" style={{ fontStyle: 'italic' }}>
+                    Type to enter a custom value. Press Enter to accept.
+                </div>
+                {props.children}
+            </components.MenuList>
+        )
+    }
+
+    const platFormNoOptions = (): string => {
```
[the remark this answers is not in the document]
Comment (vivek-devtron) [reply src/components/ciConfig/CIConfig.tsx]: Done

File: src/components/ciConfig/CIConfig.tsx (hunk 4 above)
Comment (vivek-devtron) [reply src/components/ciConfig/CIConfig.tsx]: Done

File: src/components/ciConfig/CIConfig.tsx (hunk 3 above)
Comment (vivek-devtron) [reply src/components/ciConfig/CIConfig.tsx]: Done

File: src/components/ciConfig/CIConfig.tsx (hunk 2 above)
Comment (vivek-devtron) [reply src/components/ciConfig/CIConfig.tsx]: Done

**Review by sohel-rp — APPROVED** [review]

**Review by shivani170 — APPROVED** [review]

File: src/components/ciConfig/CIConfig.tsx [hunk 7]
```diff
@@ -1,68 +1,79 @@
...
         try {
-            setLoading(true);
-            const { result } = await getCIConfig(+appId);
-            setCIConfig(result);
-            setLoading(false);
-            respondOnSuccess();
+            setLoading(true)
+            const { result } = await getCIConfig(+appId)
+            setCIConfig(result)
+            setLoading(false)
```
Comment (shivani170) [inline src/components/ciConfig/CIConfig.tsx:68]: Can set loader in finally

File: src/components/ciConfig/CIConfig.tsx [hunk 8]
```diff
@@ -124,40 +135,74 @@ function Form({ dockerRegistries, sourceConfig, ciConfig, reload, appId }) {
...
+    let _customTargetPlatorm = false
+    if (ciConfig?.dockerBuildConfig?.targetPlatform) {
+        _selectedPlatforms = ciConfig.dockerBuildConfig.targetPlatform.split(',').map((platformValue) => {
+            _customTargetPlatorm = _customTargetPlatorm || !targetPlatformMap.get(platformValue)
+            return { label: platformValue, value: platformValue }
+        })
+    }
+    const [selectedTargetPlatforms, setSelectedTargetPlatforms] = useState<OptionType[]>(_selectedPlatforms)
+    const [showCustomPlatformWarning, setShowCustomPlatformWarning] = useState<boolean>(_customTargetPlatorm)
+    const [showCustmPlatformConfirmation, setShowCustmPlatformConfirmation] = useState<boolean>(false)
```
Comment (shivani170) [inline src/components/ciConfig/CIConfig.tsx:160]: Typo `showCustomPlatformConfirmation`

Comment (vivek-devtron) [reply src/components/ciConfig/CIConfig.tsx]: Done

File: src/components/ciConfig/CIConfig.tsx (hunk 7 above)
Comment (vivek-devtron) [reply src/components/ciConfig/CIConfig.tsx:68]: Done

**Review by sohel-rp — APPROVED** [review]

File: src/components/ciConfig/CIConfig.scss [hunk 9]
```diff
@@ -72,6 +72,11 @@
             padding-left: 10px;
         }
     }
+    .target-platform__select__multi-value__label {
```
[the remark this answers is not in the document]
Comment (vivek-devtron) [reply src/components/ciConfig/CIConfig.scss:75]: It's a dynamic component

**Review by shivani170 — APPROVED** [review]

**Pull request merged**